File: Sort List/SortList.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null)
			return head;
		// count total number of elements
		int count = 0;
		ListNode p = head;
		while (p != null) {
			count++;
			p = p.next;
		}
 
		// break up to two list
		int middle = count / 2;
 
		ListNode l = head, r = null;
		ListNode p2 = head;
		int countHalf = 0;
		while (p2 != null) {
			countHalf++;
			ListNode next = p2.next;
 
			if (countHalf == middle) {
				p2.next = null;
				r = next;
			}
			p2 = next;
		}
 
		// now we have two parts l and r, recursively sort them
		ListNode h1 = sortList(l);
		ListNode h2 = sortList(r);
 
		// merge together
		ListNode merged = merge(h1, h2);
 
		return merged;
	}
    public ListNode merge(ListNode l, ListNode r){
        ListNode lp = l, rp = r;
        ListNode newhead = new ListNode(-1);
        ListNode cur = newhead;

        while(lp!=null || rp!=null){
            if(lp==null){
                cur.next = rp;
                break;
            }else if(rp==null){
                cur.next = lp;
                break;
            }else{
                if(lp.val <= rp.val){
                    cur.next = lp;
                    lp = lp.next;
                    cur = cur.next;
                }else {
                    cur.next = rp;
                    rp = rp.next;
                    cur = cur.next;
                }
            }
        }
        return newhead.next;
    }
}

File: Sort List/SortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
def merge(left, right):
    l=left
    r=right
    fakehead=ListNode(8888)
    newtemp=fakehead
    while l!=None or r!=None:
        if l==None:
            newtemp.next=r
            break
        elif r==None:
            newtemp.next=l
            break            
        else:
            if l.val<=r.val:
                newtemp.next=l
                l=l.next
                newtemp=newtemp.next
            else:
                newtemp.next=r
                r=r.next
                newtemp=newtemp.next
    return fakehead.next
class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        if head==None:
            return head
        if head.next==None:
            return head
        count=0
        temp=head
        while temp!=None:
            temp=temp.next
            count=count+1
        middle=count/2
        temp2=head
        left=head
        right=None
        countHalf=0
        while temp2!=None:
            countHalf=countHalf+1
            future=temp2.next
            if countHalf==middle:
                temp2.next=None
                right=future
                break
            temp2=future
        return merge(Solution.sortList(self,left), Solution.sortList(self,right))
        
        
